docs/content/contrib-adding-builtin-functions.md

```diff
@@ -0,0 +1,156 @@
+    if err != nil {
+        return err
+    }
+
+    // Get the first argument as an int, returning an error if it's not the correct type or not a positive value.
+    count, err := builtins.IntOperand(operands[1].Value, 2)
+    if err != nil {
+        return err
+    } else if count < 0 {
+        // Defensive check, strings.Repeat(...) will panic for count<0
+        return errors.New("count must be a positive integer")
+    }
+
+    // Return a string by invoking the given iterator function
+    return iter(ast.StringTerm(strings.Repeat(string(str), count)))
+}
+
+func init() {
+    RegisterBuiltinFunc(ast.Repeat.Name, builtinRepeat)
+}
+```
+
+In the above code, `builtinRepeat` implements the `topdown.BuiltinFunc` function type. 
+The call to `RegisterBuiltinFunc(...)` in `init()` adds the built-in function to the evaluation engine; binding the implementation to `ast.Repeat` that was registered in [an earlier step](#register-the-function).
+
+### Test
+
+All built-in function implementations must include a test suite. 
+Test cases for built-in functions are written in YAML and located under `test/cases/testdata`.
+
+We create two new test cases (one positive, expecting a string output; and one negative, expecting an error) for our built-in function:
+
+```yaml
+cases:
+  - note: repeat/positive
+    query: data.test.p = x
+    modules:
+      - |
+        package test
+        
+        p = repeated {
+          repeated := repeat(input.str, input.count)
+        }
+    input: {"str": "Foo", "count": 3}
+    want_result:
+      - x: FooFooFoo
+  - note: repeat/negative
+    query: data.test.p = x
+    modules:
+      - |
+        package test
+
+        p = repeated {
+          repeated := repeat(input.str, input.count)
+        }
+    input: { "str": "Foo", "count": -3 }
+    strict_error: true
+    want_error_code: eval_builtin_error
+    want_error: 'repeat: count must be a positive integer'
+```
+
+The above test cases can be run separate from all other tests through: `go test ./topdown -v -run 'TestRego/repeat'`
+
+See [test/cases/testdata/helloworld](https://github.com/open-policy-agent/opa/blob/main/test/cases/testdata/helloworld)
+for a more detailed example of how to implement tests for your built-in functions.
+
+> Note: We can manually test our new built-in function by [building](../contrib-development#getting-started) 
+> and running the `eval` command. E.g.: `$./opa_<OS>_<ARCH> eval 'repeat("Foo", 3)'`
+
+### Document
+
+All built-in functions must be documented in `docs/content/policy-reference.md` under an appropriate subsection.
+
+For this example, we add an entry for our new function under the `Strings` section:
+
+```markdown
+### Strings
+
+| Built-in | Description | Wasm Support |
+| ------- |-------------|---------------|
+...
+| <span class="opa-keep-it-together">``output := repeat(string, count)``</span> | ``output`` is ``string`` repeated ``count``times | ✅ |
```

Comment:
Haven't seen any wasm mention here, so I think this should say "SDK dependent"

Reply:
Changing to `SDK-dependent`. Would probably be good to also add info to this how-to about Wasm. I don't think I can currently bild and test for that on my M1 mac; so something to return to later, perhaps, either for me or someone else.

Reply:
It's cumbersome, but you could perhaps use a docker container...? Not sure how that's done on M1.